### PR TITLE
[dagster-pipes] add `PipesThreadedMessageReader`

### PR DESCRIPTION
**Pull request by danielgafni**
## Summary & Motivation

Sometimes we can only start readings messages (and logs) after the external process has been launched. 

This PR:
- Moves most of `PipesBlobStoreMessageReader`'s functionality into `PipesThreadedMessageReader`
- practically renames `PipesBlobStoreMessageReader` to `PipesThreadedMessageReader` with very little refactoring
- adds `can_read` method. For `PipesBlobStoreMessageReader` it defaults to `True` since it was only used with known locations until now. 

Resolve [FEAT-755](https://linear.app/dagster-labs/issue/FEAT-755/support-reading-pipes-messages-and-logs-from-locations-known-at)

## Changelog

Insert changelog entry or "NOCHANGELOG" here.

- [x] `NEW` _(introduced a base `ThreadedPipesMessageReader`. It supports reading messages and logs from various sources concurrently. Supports reading from locations only known after the external process has been launched)_

### CODE REVIEW

**Comment by danielgafni** [comment]
* **#24469** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24469?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @danielgafni and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (danielgafni) [inline python_modules/dagster/dagster/_core/pipes/utils.py:341]: I'm now sure if we actually need this event, seems like checking for `self.target_is_readable` is enough

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (schrockn) [inline python_modules/dagster/dagster/_core/pipes/utils.py:251]: would prefer to just canonicalize to one representation or the other

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (schrockn) [inline python_modules/dagster/dagster/_core/pipes/utils.py:506]: `state` is worthy of commentary here. It is very generic so difficult to understand its intended use.

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (schrockn) [inline python_modules/dagster/dagster/_core/pipes/utils.py:338]: Generally prefer to avoid this sort of mutability if possible.

**Comment by schrockn** [comment]
@smackesey I believe has the most context here?

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (danielgafni) [reply python_modules/dagster/dagster/_core/pipes/utils.py:338]: Should we store these provided params in a separate attribute? 
The whole point is to be able to receive additional params provided manually.

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (danielgafni) [reply python_modules/dagster/dagster/_core/pipes/utils.py:265]: This means changing all current implementations including guides and tutorials, also a breaking change. 

Can we do it?

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (danielgafni) [inline python_modules/dagster/dagster/_core/pipes/utils.py:367]: not sure what's a good new name for this method

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (danielgafni) [reply python_modules/dagster/dagster/_core/pipes/utils.py:338]: I'm working on an example of `update_params` usage with `PipesCloudWatchMessageReader` as a separate PR

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (danielgafni) [reply python_modules/dagster/dagster/_core/pipes/utils.py:557]: I think it's actually not needed here. I'll remove it.

**Comment by danielgafni** [comment]
I have another idea... seems like it's possible to simplify this code even more by reusing the same general `PipesLogReader` for both reading messages and logs. 

Actually, the `_message_reader` thread here does pretty much the same as the `_logs_reader` threads. 

The only difference is where are we sending the data. 

There are 3 possible destinations and 2 filtering modes: 
1. drop Dagster messages and route to `stdout`
2. drop Dagster messages and route to `stderr`
3. keep only Dagster messages and route to message handler (Dagster event log)

Therefore, looks like `ThreadedMessageReader` can be refactored to receive: 
- a main `PipesLogReader` for messages/event. Internally we can call it with `filtering="messages", destination="dagster"`. 
- multiple `PipesLogReader` for logs. Internally we can call it with `filtering="logs". Each can have `destination=sys.stdout/sys.stderr` set. 

This way we won't have to be writing the implementation of reading messages/logs twice. 

Are there any reasons not to do it?

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (schrockn) [reply python_modules/dagster/dagster/_core/pipes/utils.py:253]: I mean this proposes renaming the class right? Seems like we are crossing the boundary of breaking change anyways.

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (schrockn) [inline python_modules/dagster/dagster/_core/pipes/utils.py:281]: This is worth of a doc string. What does "readable" mean in this context? When is something not readable?

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (schrockn) [inline python_modules/dagster/dagster/_core/pipes/utils.py:240]: Very obtuse name. `TCursor` or something would be more helpful.

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (schrockn) [inline python_modules/dagster/dagster/_core/pipes/utils.py:326]: can you link me to an actual usage of this? Would be very clarifying.

**Review by schrockn — CHANGES_REQUESTED** [review]
Popping back into your queue

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (danielgafni) [reply python_modules/dagster/dagster/_core/pipes/utils.py]: Renamed to `can_read`. 

`can_read` should return `False` if: 
- `MessageReader` is missing some params which are expected to be submitted once known with `.update_params`
- perahs the location to read from is not physically available yet

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (danielgafni) [reply python_modules/dagster/dagster/_core/pipes/utils.py]: I added usages with `PipesCloudWatchMessageReader` inside `PipesGlueClient` and `PipesEMRServerlessClient` in the same PR. Still to be implemented for `PipesECSClient`.

File: python_modules/libraries/dagster-aws/dagster_aws/pipes/clients/emr_serverless.py
Comment (danielgafni) [inline python_modules/libraries/dagster-aws/dagster_aws/pipes/clients/emr_serverless.py:302]: @schrockn here is an example of calling `.add_log_reader`. 

`log_stream` is only known after the external process has been launched so we have to manually add it here.

File: python_modules/libraries/dagster-aws/dagster_aws/pipes/clients/glue.py
Comment (danielgafni) [inline python_modules/libraries/dagster-aws/dagster_aws/pipes/clients/glue.py:225]: @schrockn another example

**Comment by danielgafni** [comment]
Please take a look at the new iteration @schrockn @smackesey. Key changes: 

- renamed `target_is_readable` to `can_read`
- `update_params` only updates params for the `PipesMessageReader` itself. It doesn't affect LogReaders. 
- Instead, there is a `add_log_reader` method which can be used to submit a new `PipesLogReader`. `PipesMessageReader` will then start this LogReader when possible. This workflow better fits Pipes usage, in particular, it doesn't require the user to initialize all possible LogReaders from the start (which can become really complicated, ECS is a good example).

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (danielgafni) [reply python_modules/dagster/dagster/_core/pipes/utils.py]: Not really renaming. 

99% of the old class functionality is now in a new class. 
The old class now simply inherits from the new class and customizes the remaining 1%. 

What I ended up doing is: 
- update the type annotation 
- but actually still support the old-style list internally so there are no runtime failures

**Comment by danielgafni** [comment]
Got `PipesThreadedMessageReader` and the new `PipesCloudWatchMessageReader` (now a child class) working! 
`PipesCloudWatchMessageReader` can now read messages & logs in background. 

Still have to port `ECSPipesClient` to use the new `PipesCloudWatchMessageReader`.

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (smackesey) [inline python_modules/dagster/dagster/_core/pipes/utils.py:255]: this can just be `log_readers or {}`

**Comment by smackesey** [comment]
I am having difficulty grasping the value prop here-- that's not to imply this isn't worthwhile, just that it's not obvious to me from the description and diff. So a few questions:

This seems like an increase in complexity over the status quo. What new use case is this additional complexity supporting? Is it using a message reader in cases where for some reason we cannot read messages prior to the launch of the external process? What is an example of this situation?

>adds add_log_reader method which can be used to submit PipesLogReader for targets only known after the external process has been launched. The PipesThreadedMessageReader will start these added PipesLogReader once possible.

Similarly, what is an example of a scenario where we need this? In some sense we already support log readers for targets only known after the external process has launched-- the `PipesDbfsLogReader`, for instance, needs to receive the cluster id (via the `opened` message) before the target path can be resolved. What isn't supported is a dynamic _number_ of logs-- is that what you're trying to add support for here?

**Comment by danielgafni** [comment]
Hey @smackesey, thanks for the review. 

> Is it using a message reader in cases where for some reason we cannot read messages prior to the launch of the external process? What is an example of this situation?

Yes, there are examples like this:
- AWS EMR [some docs](https://aws.github.io/aws-emr-containers-best-practices/troubleshooting/docs/where-to-look-for-spark-logs/#log-location-of-jobrunner-driver-executor-in-s3)
- AWS ECS, AWS Glue, basically any services which creates a random ID for the job
 
Actually, this is not a critical issue since we always can use an escape hatch such as using a non-default `message_writer` (like `PipesS3MessageWriter`). But doing this is an additional step for the user. I think we should try to be utilizing a "native" channel for passing messages before asking the user to configure a non-default `message_writer`. In most cases that's stdout/stderr.  

> What isn't supported is a dynamic number of logs-- is that what you're trying to add support for here?

This is a bigger issue. Services like AWS ECS or AWS EMR can be writing multiple log files (or CloudWatch streams). And the locations might be unclear before launching the job, because often AWS would merge some "base" (cluster) configuration with job-specific overrides, and then there might be some interactions in your job parameters which can enable or disable specific log targets. 

This logic would be very hard to predict on our side before creating the MessageReader. It's much more simple and reliable to:
1. Launch the job
2. Describe the job
3. Create a `LogReader` per each log target 

You can look at just the ECS in this PR https://github.com/dagster-io/dagster/pull/24809 to get an idea about what I'm trying to do.

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (danielgafni) [reply python_modules/dagster/dagster/_core/pipes/utils.py:255]: Because `self.log_readers` must be a `Dict` (so we can assign to keys later on), while the constructor argument accepts a `Mapping` (a more general type).

Comment (smackesey) [reply python_modules/dagster/dagster/_core/pipes/utils.py:255]: Aren't the new keys assigned in `extra_log_readers`

Comment (danielgafni) [reply python_modules/dagster/dagster/_core/pipes/utils.py:255]: They are initially assigned there but eventually get moved to `self.log_readers` once started. So we can have them all in one place. 

I'm up to changing this detail, it's not really important, but that's the current implementation.

**Comment by smackesey** [comment]
OK I've now looked over this PR and the upstack Cloudwatch one in more detail. I agree that configuring the message reader using information available only post-external-process-launch is a worthwhile goal, but I'm hesitant to move forward with the approach in this PR.

This is introducing significant mutability into the pipes system. The new `PipesThreadedMessageReader` has two-stage initialization. There is object construction and then the call to `update_params`, which updates a separate `state` object and sets the `self.runtime_params_updated` flag.

This temptation to introduce these "multi-phase" objects is something that came up a lot during initial pipes development-- it comes with the problem space due to the differing stages of available information. I did this in several of my own prototype PRs and remember @schrockn flagging it. I think when I was writing the code it felt natural, but it ends up making the system more difficult to reason about and debug.

An example of this is `PipesMessageWriter`/`PipesMessageWriterChannel` in the external process. The `PipesMessageWriter` needs to be passed into `open_dagster_pipes`. It needs to be constructed by and arbitrarily parameterized by the user. But, to do its work it needs to receive the params payload from the orchestration process, which is only available after the `open_dagster_pipes` call. Some initial designs for `PipesMessageWriter` included a second stage of initialization, after receiving the params, analogous to setting `PipesThreadedMessageReader.runtime_params_updated` in this PR. But instead we introduced `PipesMessagesWriterChannel`, a new object generated by the message writer once the params were received.

I think that was the right decision and I'm wondering whether we can do something similar here. Would it be possible to introduce `PipesMessageReaderChannel` that is scoped to the new phase of the lifecycle (post-external-process-launch on orchestration side) being introduced in this PR? You've uncovered an important gap in Pipes functionality, but to fill that gap we should consider a new object scoped to a particular phase of the pipes lifecycle instead of cramming complexity into the message reader.

**Comment by danielgafni** [comment]
I'm open to this, but could you perhaps elaborate on what would the `PipesMessageReaderChannel` be doing exactly?

**Comment by smackesey** [comment]
>I'm open to this, but could you perhaps elaborate on what would the PipesMessageReaderChannel be doing exactly?

Forget about `PipesMessageReaderChannel` for now-- let's zoom out. I think what is concerning in the current PR is the introduction of a new lifecycle state that's not explicitly modeled.

This is the current pipes lifecycle ("OP"/"EP" for orchestration/external process). I've called out where this new stage is:

- [OP] `PipesContextInjector` and `PipesMessageReader` are initialized
- [OP] `open_pipes_session` called
    - [OP] `PipesMessageHandler` constructed, wraps `PipesMessageReader`
    - [OP] `PipesContextInjector.inject_context()` called, yields params
    - [OP] `PipesMessageHandler.handle_messages()` called; this calls wrapped 
    - [OP]`PipesMessageReader.read_messages()` and yields the params
- [OP] `open_pipes_session` yields `PipesSession` object, which wraps the yielded context and message params
- [OP] user/client code launches external process, passing context/message param pulled off the `PipesSession`
- [OP] !!!!!!!!!!!!!!!!!!!!! NEW STAGE !!!!!!!!!!!!!!!!!!!!!!!!!
- [EP] `PipesParamsLoader`, `PipesContextLoader`, `PipesMessageWriter` constructed
- [EP] `open_dagster_pipes` called
- [EP] `PipesContext` constructed
	- [EP] `PipesParamsLoader.load_{context,messages}_params` called
	- [EP] `PipesContextLoader.load_context(context_params)` called, yields `PipesContextData`
	- [EP] `PipesMessageWriter.open(messages_params)` called, yields `PipesMessageChannel`
	- [EP] `PipesMessageChannel` writes "opened" message
- [OP] `PipesMessageReader.on_opened()` called with opened payload 

Rather than modeling this stage within the whole pipes lifecycle, this PR is _ad hoc_ modifying a specific `PipesMessageReader` subclass to sort of hack this stage in. It does this by waiting for user code to call `PipesThreadedMessageReader.update_params`, which then updates a nebulous `state` object that exists only on `PipesThreadedMessageReader`.

This feels like it has too many degrees of freedom. Consider that `PipesMessageReader` as currently implemented does something similar already with the "opened" payload, but it is more constrained. "opened" could have been implemented by having a generic `state` object on `PipesMessageReader` that is modified by a new `update_params` function, but instead we have an `on_opened` hook that is called every pipes session and that sets the `opened_payload` attribute instead of modifying a generic `state`.

I think we can do something similar here. Unfortunately we're not going to be able to trigger a `PipesMessageReader.on_launched()` hook as part of every session, because launch is under the control of user code. But, I think we can add this hook, optionally called by the user/client, to `PipesMessageReader` and have it set a `launched_payload` object analogous to `opened_payload`. Individual `PipesMessageReader.read_messages` routines could then condition behavior on the `launched_payload` instead of `state`.

I think it would be clarifying if you could add this `on_launched()` hook in a downstack PR and modify the present PR to use that in lieu of `update_params()`/`state`.

**Comment by danielgafni** [comment]
Alright, all of this makes sense. 

I didn't want to touch the base PipesMessageReader on purpose (wasn't sure if the proposed changes are general enough), but now I can see how it would be useful to create this hook there instead.

**Comment by danielgafni** [comment]
Separate question: currently we only launch `PipesLogReader` after the `PipesMessageReader` starts. 
But sometimes there can be logs available for reading even if the Pipes process has failed to start and there are no messages available, and the `PipesMessageReader` can't start. 

Example: I'm setting up `AWS EMR` pipes. The remote process failes to start, but EMR creates logs with startup errors. 

It would be great if we were able to forward these logs to Dagster even if the main Pipes remote process was not launched properly.

**Comment by smackesey** [comment]
>Separate question: currently we only launch PipesLogReader after the PipesMessageReader starts. But sometimes there can be logs available for reading even if the Pipes process has failed to start and there are no messages available, and the PipesMessageReader can't start.

Well, the "start" of the `PipesMessageReader` is not well-defined. There are only two framework-level well-defined stages:

- `PipesMessageReader.read_messages()` is called. This happens inside the call to `open_pipes_session()`, so if this is the "start" of the message reader, then it isn't blocked by any failure to start the process.
- After the message reader has received the opened payload. In this is what is meant by "start", then yes, failing to start the external process will fail to start the message reader.

The only place that `PipesLogReader` is currently used, I think, is in `PipesBlobStoreMessageReader`, and in that case yes, we only start the log readers after we have received the opened payload. But that's a detail of `PipesBlobStoreMessageReader`-- it's not enforced by `PipesLogReader` itself.

>To enable this, we could add a similar on_opened lifecycle hook to PipesLogReader (so it can receive extra params after initialization) and call it inside Pipes automatically.

I'm assuming you mean `on_launched` here instead of `on_opened`, because the `PipesLogReader` is already using the `opened_payload` in `PipesBlobStoreMessageReader`. If that's the case, I don't think we need the hook-- you could for instance adjust `logs_thread` in `PipesBlobStoreMessageReader` to wait for `launched_payload` instead of opened payload if that's when you wanted to start reading.

**Comment by danielgafni** [comment]
Sorry I've meant on_launched, right. 

Ok, just wanted to make sure what I'm proposing makes sense. I'll edit PipesThreadedMwssageReader to start the PipesLogReaders as soon as possible.

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (smackesey) [inline python_modules/dagster/dagster/_core/pipes/utils.py:355]: Having nested `can_start` checks feels off to me. Why not remove this outer one and just wait for the launch payload?

**Review by smackesey — COMMENTED** [review]
So this PR is pretty hard to review because there are multiple significant indepedent design modifications being made:

- Introducing launch triggering to the start of message reading
- Adding "cursor"-based reading
- Adding mutable number of logs.

I think it would be useful for these to be broken up so that they can be independently evaluated/discussed/merged. Probably just adding the launch-payload and `can_start` check on `PipesBlobStoreMessageReader` can land pretty quick, but I need to see the mutable logs proposal in isolation. And I don't understand why the generic cursor-based mechanism is also being added in this PR.

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (danielgafni) [reply python_modules/dagster/dagster/_core/pipes/utils.py:355]: Because the starting condition is individual for each `PipesLogReader`. We can't assume they all will become startable at the same time. 

For example, they might be waiting for different objects to appear on S3, and thes objects are uploaded by an external system one by one.

**Comment by danielgafni** [comment]
I moved dynamic `PipesLogReader` logic into https://github.com/dagster-io/dagster/pull/25066

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (smackesey) [inline python_modules/dagster/dagster/_core/pipes/utils.py:311]: We need a better name than `can_start` since it's not obvious what "start" means (e.g. the threads are already running when we are checking this). Maybe `are_messages_readable`.

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (smackesey) [inline python_modules/dagster/dagster/_core/pipes/utils.py:522]: cursor: Any should be `cursor: TCursor` right

Comment (smackesey) [reply python_modules/dagster/dagster/_core/pipes/utils.py:522]: Also please add more of an explanatory comment about `download_messages_chunk` and `download_messages_parts`. Very confusing for anyone fresh to the code to see both methods here.

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (smackesey) [reply python_modules/dagster/dagster/_core/pipes/utils.py:324]: How about just `download_messages`

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (smackesey) [inline python_modules/dagster/dagster/_core/pipes/utils.py:245]: Needs a docstring

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (smackesey) [inline python_modules/dagster/dagster/_core/pipes/utils.py:247]: `counter` is a feature of the blob store reader but not this one right

**Review by smackesey — CHANGES_REQUESTED** [review]

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (danielgafni) [reply python_modules/dagster/dagster/_core/pipes/utils.py:363]: The problem is, the launched_payload is not guaranteed to be received. 
So we can't have a separate waiting loop just for it. 

What we can do is make a variable like `payload_received` and update it in the loop.

**Comment by smackesey** [comment]
Per offline discussion let's add some tests

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (smackesey) [inline python_modules/dagster/dagster/_core/pipes/utils.py:246]: Since there is no actual class called "MessageReader", prefer:

"""A base class for message readers that read messages and logs in background threads."""

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (smackesey) [inline python_modules/dagster/dagster/_core/pipes/utils.py:250]: `log_readers` type should match the annotation in `__init__`

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (smackesey) [inline python_modules/dagster/dagster/_core/pipes/utils.py:339]:
```
cursor (Optional[Any]): Cursor specifying start location from which to download 
    messages in a stream. The format of the value varies with the message reader 
    implementation. It might be an integer index for a line in a log file, or a 
    timestamp for a message in a time-indexed stream.
```

**Review by smackesey — APPROVED** [review]
I've made some final requests but I'm approving-- please implement requested changes to your best judgment and land. It is possible we will tweak implementation in a follow-up but it will be useful to land a version of this now to maintain momentum.

**Comment by smackesey** [comment]
Also I think you can just dismiss @schrockn requested changes from two weeks ago, it's out of date and shouldn't block landing this.

File: python_modules/dagster/dagster/_core/pipes/utils.py
Comment (danielgafni) [reply python_modules/dagster/dagster/_core/pipes/utils.py:372]: Oh I had an opposite impression that we agreed to make this more future-proof in case we are going to have multiple parameter updates... 

But I don't mind, it's also ok to do as you proposed.

Comment (smackesey) [reply python_modules/dagster/dagster/_core/pipes/utils.py:372]: ok, let's go ahead and land

**Pull request merged**
